File: leetcode/python/0049-group-anagrams.py
# path: leeetcode\python\0049-group-anagrams.py
# link: https://leetcode.com/problems/group-anagrams/
# 2023-04-10
# leetcode: Medium

from collections import defaultdict
from typing import List

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # 0(m * n)
        res = defaultdict(list)

        for s in strs:
            count = [0] * 26

            for c in s:
                count[ord(c) - ord("a")] += 1

            res[tuple(count)].append(s)
            
        return res.values()

File: leetcode/typescript/0049-group-anagrams.ts
/**
 * Problem: 49. Group Anagrams
 * Time O(N) | Space O(N)
 * Path: leetcode\typescript\0049-group-anagrams.ts
 * Link: https://leetcode.com/problems/group-anagrams/
 * Leetcode: Medium
 * Date: 2023-05-05
 * Redoing all the Medium problems from scratch to get used to TypeScript
 */

function groupAnagrams(strs: string[]): string[][] {
	const map: Map<string, Array<string>> = new Map();

	for (const str of strs) {
		const sortedStr = str.split("").sort().join("");
		if (map.has(sortedStr)) {
			map.get(sortedStr)?.push(str);
		} else {
			map.set(sortedStr, [str]);
		}
	}

	return Array.from(map.values());
}
